src/main/java/com/wavefront/sdk/grpc/WavefrontClientInterceptor.java

```diff
@@ -0,0 +1,451 @@
+      // update requests inflight
+      getGaugeValue(new MetricName(REQUEST_PREFIX + methodName + ".inflight", allTags)).
+          decrementAndGet();
+      getGaugeValue(new MetricName("client.total_requests.inflight",
+          overallAggregatedPerSourceTags)).decrementAndGet();
+      // rpc latency
+      wfGrpcReporter.updateHistogram(new MetricName(
+          RESPONSE_PREFIX + methodWithStatus + ".latency", histogramAllTags), rpcLatency);
+      wfGrpcReporter.incrementCounter(new MetricName(RESPONSE_PREFIX + methodWithStatus +
+          ".total_time", allTags), rpcLatency);
+      // request, response size
+      wfGrpcReporter.updateHistogram(new MetricName(
+          REQUEST_PREFIX + methodName + ".bytes", histogramAllTags), requestBytes.get());
+      wfGrpcReporter.updateHistogram(new MetricName(
+          RESPONSE_PREFIX + methodName + ".bytes", histogramAllTags), responseBytes.get());
+      wfGrpcReporter.incrementCounter(new MetricName(REQUEST_PREFIX + methodName + ".total_bytes",
+          allTags), requestBytes.get());
+      wfGrpcReporter.incrementCounter(new MetricName(RESPONSE_PREFIX + methodName + ".total_bytes",
+          allTags), responseBytes.get());
+      // streaming stats
+      if (streamingStats) {
+        wfGrpcReporter.updateHistogram(new MetricName(
+                REQUEST_PREFIX + methodName + ".streaming.messages_per_rpc", histogramAllTags),
+            requestMessageCount.get());
+        wfGrpcReporter.updateHistogram(new MetricName(
+                RESPONSE_PREFIX + methodName + ".streaming.messages_per_rpc", histogramAllTags),
+            responseMessageCount.get());
+      }
+      Map<String, String> aggregatedPerShardTags = new HashMap<String, String>() {{
+        put(CLUSTER_TAG_KEY, applicationTags.getCluster() == null ? NULL_TAG_VAL :
+            applicationTags.getCluster());
+        put(SERVICE_TAG_KEY, applicationTags.getService());
+        put(SHARD_TAG_KEY, applicationTags.getShard() == null ? NULL_TAG_VAL :
+            applicationTags.getShard());
+        put(GRPC_SERVICE_TAG_KEY, grpcService);
+        put(SOURCE_KEY, WAVEFRONT_PROVIDED_SOURCE);
+      }};
+      Map<String, String> aggregatedPerServiceTags = new HashMap<String, String>() {{
+        put(CLUSTER_TAG_KEY, applicationTags.getCluster() == null ? NULL_TAG_VAL :
+            applicationTags.getCluster());
+        put(SERVICE_TAG_KEY, applicationTags.getService());
+        put(GRPC_SERVICE_TAG_KEY, grpcService);
+        put(SOURCE_KEY, WAVEFRONT_PROVIDED_SOURCE);
+      }};
+      Map<String, String> aggregatedPerClutserTags = new HashMap<String, String>() {{
+        put(CLUSTER_TAG_KEY, applicationTags.getCluster() == null ? NULL_TAG_VAL :
+            applicationTags.getCluster());
+        put(GRPC_SERVICE_TAG_KEY, grpcService);
+        put(SOURCE_KEY, WAVEFRONT_PROVIDED_SOURCE);
+      }};
+      Map<String, String> aggergatedPerApplicationTags = new HashMap<String, String>() {{
```

Comment:
Reusing ImmutableMap.Builder instance here would make the code more concise and readable.